lib/ytdl.py

```diff
@@ -18,34 +38,310 @@
+                    "source!"
+                )
+                self.current_source_index += 1
+                return self.read()
+            else:
+                # return silent audio data until next source is ready
+                return AudioSegment.silent(duration=20, frame_rate=48000).raw_data
+
+        # calculate volume for fade in/out effect
+        factor = min([time_elapsed, time_remaining])
+        source.volume = (factor / (self.audio_fade_time * 1000)) * self.volume
+
+        return source.read()
 
 
 class YTDLSource(discord.PCMVolumeTransformer):
-    def __init__(self, source, *, data, volume=1.0):
+    """Wrapper over `discord.PCMVolumeTransformer` for the purpose of keeping audio data
+    in memory. This gives us the option to download the file while streaming it to the
+    client at the same time, without needing to fetch from YouTube twice.
+    """
+
+    # TODO: programmed this in mind that download functionality will always be used,
+    # but should check to make it still works even without it
+    # NOTE: because of the way we built this, it'll also be possible to detect any
+    # issues with the stream before the bytes are actually getting read, so we have a
+    # chance to repair audio streams and continue downloading if we only get like half
+    # way through a download and an error occurs
+    def __init__(
+        self,
+        source: discord.FFmpegPCMAudio,
+        *,
+        volume=1.0,
+        loop: asyncio.AbstractEventLoop = None,
+        metadata: Mapping[str, Any] = None,
+        download=False,
+    ):
         super().__init__(source, volume)
-        self.data = data
-        self.title = data.get("title")
-        self.url = data.get("url")
+        self.source = source
+        self.loop = loop or asyncio.get_event_loop()
+        self.metadata = metadata
+        self.is_stream_ready = False
+        self.is_download_ready = False
+        self._audio_buffer: List[bytes] = []
+        self._audio_buffer_index = -1
+
+        if download:
+            self.download_task = self.loop.create_task(self.start_download())
 
     @classmethod
-    async def from_url(cls, url, *, loop=None, stream=False):
-        loop = loop or asyncio.get_event_loop()
+    async def from_query(
+        cls, query: str, *, loop: asyncio.AbstractEventLoop = None, download=True
+    ):
+        """Instantiate `YTDLSource` from a generic query or URL."""
+        logger.debug(f"Fetching metadata for query: {query}")
+        start_time = perf_counter()
 
+        ytdl = youtube_dl.YoutubeDL(YTDL_FORMAT_OPTIONS)
+        loop = loop or asyncio.get_event_loop()
         data = await loop.run_in_executor(
-            None, lambda: ytdl.extract_info(url, download=not stream)
+            None, lambda: ytdl.extract_info(query, download=False)
         )
-
         if "entries" in data:
             data = data["entries"][0]
```

Comment:
I'll need to review and double-check what this line is actually doing. Currently, there's an issue with playlists taking forever to load, and when they finally do, they only play the first song in the playlist it seems, and I think this line might have something to with that.

Reply:
Turns out I was right about this one. When the youtube extractor extracts a playlist it inserts them into `data.entries`. Made adjustments to support playlists in the latest commit